src/lib/descriptor/mod.rs

```diff
@@ -14,69 +14,175 @@ mod cargo_alias;
+};
+use crate::{io, scriptengine, version};
+
+static RE_VARIABLE: Lazy<Regex> = Lazy::new(|| Regex::new(r"\$\{.*}").unwrap());
+
+fn merge_env_depends_on_extract(val: &str) -> Vec<&str> {
+    let mut depends_on = vec![];
+
+    for matched in RE_VARIABLE.find_iter(val) {
+        let matched = matched.as_str().trim();
+        // remove the first two characters (`${`)
+        let (_, matched) = matched.split_at(2);
+        // remove the last character (`}`)
+        let (matched, _) = matched.split_at(matched.len() - 1);
+
+        depends_on.push(matched);
+    }
+
+    depends_on
+}
+
+fn merge_env_unique<'a>(
+    vals: &'a [&'a IndexMap<String, EnvValue>],
+) -> Vec<(&'a str, &'a EnvValue)> {
+    let mut visited = HashSet::new();
+    let mut unique = vec![];
+
+    // iterate through the list in reverse, only taking the first value, then
+    // reversing again to make sure that we still adhere to the order.
+    // This way we will only ever take the latest value.
+    for (key, val) in vals.iter().map(|map| map.iter()).flatten().rev() {
+        if visited.contains(&key.as_str()) {
+            continue;
+        }
+
+        visited.insert(key.as_str());
+        unique.push((key.as_str(), val))
+    }
+
+    unique.reverse();
+    unique
+}
+
+fn merge_env_depends_on(val: &EnvValue) -> Vec<&str> {
+    match val {
+        EnvValue::Value(value) => merge_env_depends_on_extract(value),
+        EnvValue::Decode(EnvValueDecode { source, .. }) => merge_env_depends_on_extract(source),
+        EnvValue::List(values) => values
+            .iter()
+            .map(|value| merge_env_depends_on_extract(value))
+            .reduce(|mut acc, mut other| {
+                acc.append(&mut other);
+                acc
+            })
+            .unwrap_or_default(),
+        EnvValue::Conditional(EnvValueConditioned { value, .. }) => {
+            merge_env_depends_on_extract(value)
+        }
+        EnvValue::PathGlob(EnvValuePathGlob { glob, .. }) => merge_env_depends_on_extract(glob),
+        _ => vec![],
+    }
+}
 
 fn merge_env(
-    base: &mut IndexMap<String, EnvValue>,
-    extended: &mut IndexMap<String, EnvValue>,
-) -> IndexMap<String, EnvValue> {
-    let mut merged = IndexMap::<String, EnvValue>::new();
+    base: &IndexMap<String, EnvValue>,
+    ext: &IndexMap<String, EnvValue>,
+) -> Result<IndexMap<String, EnvValue>, String> {
+    let combined = [base, ext];
+    let combined: Vec<_> = merge_env_unique(&combined);
 
-    for (key, value) in base.iter() {
-        let key_str = key.to_string();
-        let value_clone = value.clone();
-        merged.insert(key_str, value_clone);
+    let mut graph: GraphMap<&str, (), _> = DiGraphMap::new();
+
+    let keys: HashSet<_> = combined.iter().map(|(key, _)| *key).collect();
+    for key in keys {
+        graph.add_node(key);
     }
 
-    for (key, value) in extended.iter() {
-        let key_str = key.to_string();
+    debug!("initial graph: {graph:?}");
```

Comment:
lets just make sure the min version of rust that cargo-make supports, supports this feature (i think it does...)